ray_tracing.py

```diff
@@ -111,25 +118,71 @@ def trace_ray(rayO, rayD):
     col_ray += obj.get('specular_c', specular_c) * max(np.dot(N, normalize(toL + toO)), 0) ** specular_k * color_light
     return obj, M, N, col_ray
 
-def add_sphere(position, radius, color):
+
+def add_sphere(position, radius, color, reflection=.2,
+               transparency=0.0, refraction=0.0):
     return dict(type='sphere', position=np.array(position),
-        radius=np.array(radius), color=np.array(color), reflection=.5)
+                radius=np.array(radius), color=np.array(color),
+                reflection=reflection, transparency=transparency,
+                refraction=refraction)
+
 
-def add_plane(position, normal):
+def add_plane(position, normal,
+              diffuse_c=.75, specular_c=.5, reflection=.25):
     return dict(type='plane', position=np.array(position),
-        normal=np.array(normal),
-        color=lambda M: (color_plane0
-            if (int(M[0] * 2) % 2) == (int(M[2] * 2) % 2) else color_plane1),
-        diffuse_c=.75, specular_c=.5, reflection=.25)
+                normal=np.array(normal),
+                color=lambda M: (color_plane0
+                                 if (int(M[0] * 2) % 2) == (int(M[2] * 2) % 2) else color_plane1),
+                diffuse_c=diffuse_c, specular_c=specular_c, reflection=reflection)
+
+
+def refract_ray(vect, n, coeff):
+    nv = np.dot(n, vect)
+    if nv > 0:
+        return refract_ray(vect, n * -1, 1 / coeff)
+    a = 1 / coeff
+    D = 1 - a * a * (1 - nv * nv)
+    if D < 0:
+        return None
+    b = nv * a + np.math.sqrt(D)
+    return (a * vect) - (b * n)
+
+
+def process_tracing(rayO, rayD, reflection, col, depth, refraction):
+    if depth >= depth_max:
+        return
+
+    after_trace = trace_ray(rayO, rayD)
+    if not after_trace:
+        return
+
+    obj, M, N, col_ray = after_trace
+    col += reflection * col_ray
+
+    # Reflection
+    rayO1, rayD1 = M + N * refraction * .0001, normalize(rayD - 2 * np.dot(rayD, N) * N)
+    if process_tracing(rayO1, rayD1, reflection * obj.get('reflection', 1.), col, depth + 1, refraction):
+        return
+
+    # Refraction
+    ray = refract_ray(rayD, N, obj.get('refraction', 1.))
+    if ray is None:
+        return
+    rayO2, rayD2 = M - N * refraction * .0001, ray
+    if process_tracing(rayO2, rayD2, reflection * obj.get('transparency', 1.), col, depth + 1, refraction * (-1)):
+        return
+
+    return
+
 
 # List of objects.
 color_plane0 = 1. * np.ones(3)
 color_plane1 = 0. * np.ones(3)
-scene = [add_sphere([.75, .1, 1.], .6, [0., 0., 1.]),
-         add_sphere([-.75, .1, 2.25], .6, [.5, .223, .5]),
-         add_sphere([-2.75, .1, 3.5], .6, [1., .572, .184]),
+scene = [add_sphere([.75, .1, 1.], .6, [0., 0., 1.], transparency=1, refraction=1),  # blue

```

Comment:
As I understand, the blue ball should be transparent, but on the picture we see that it is not so. Please fix it.
And also its reflection shouldn't be visible on the orange ball and on the floor too.

Reply:
Agree
Fixed in next commit